AP_Rangefinder: use offset param value in all drivers

rmackay9 opened the pull request:
This PR changes all rangefinder drivers to add the RNGFNDx_OFFSET param value to distances.  This parameter's default is 0 so hopefully it will have no effect for most users.

The change to the PWM driver fixes issue https://github.com/ArduPilot/ardupilot/issues/13252 which is a regression in Copter-4.0.0 vs Copter-3.6.12.

For consistency the offset is always added at the moment the distance from the sensor is stored in the _state.distance_cm variable.  I hope this consistency will reduce the possibility of bugs being introduced in the future.

Some things to note:

- the offset is applied before the RNGFND_MIN_CM and RNGFND_MAX_CM values are used to determine the state of the rangefinder.  I think this is correct because this is consistent with how the analog driver and [Copter-3.6.x PWM driver](https://github.com/ArduPilot/ardupilot/blob/Copter-3.6/libraries/AP_RangeFinder/AP_RangeFinder_PX4_PWM.cpp#L120) used it.
- the AP_RangeFinder_MAVLink driver is inconsistent with other drivers in that it [sets the distance to zero if there is NoData](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_RangeFinder/AP_RangeFinder_MAVLink.cpp#L71).  I have left it reporting "0" as opposed to "0+OFFSET".

WickedShell commented:
Offset should only be applied to the analog inputs. We can apply it to PWM for consistency, but it's really not correct on the others. We have a POS parameter to correct for mounting aspects. `OFFSET` is purely for sensor reading errors which is a different problem.

rmackay9 commented:
@WickedShell,

I see your point but it's got a couple of practical issues that come with it.

1. The OFFSET parameter description pretty clearly says that it is a cm offset for all non-analog drivers.  Clearly it's not and we could of course change the parameter description but this was the intention of the parameter.
2. Only the EKF uses the POS_X/Y/Z parameters and the EKF's height-above-terrain estimate is only used with optical flow (as far as I know).  Most Copter users want terrain following using a lidar which, in Copter at least, is done outside the EKF.  This leads us to two possible bits which are large scope creep

   - Switch copter over to using EKF's height-above-terrain estimate
   - Make a new height-above-terrain estimator outside the EKF and switch everyone to this

WickedShell commented:
> The OFFSET parameter description pretty clearly says that it is a cm offset for all non-analog drivers. Clearly it's not and we could of course change the parameter description but this was the intention of the parameter.

It doesn't indicate that at all for the serial ones (which is one of the ones that driving the comment from my point of view).

> Only the EKF uses the POS_X/Y/Z parameters and the EKF's height-above-terrain estimate is only used with optical flow (as far as I know).

The EKF may be the only one using it now, but it's definitely not the only one that should be using it. You actually generally speaking want to do attitude correction, particularly on larger vehicles, which needs a full 3D correction.

Basically I'm objecting to this as it's making this harder to handle correctly in the future, unless someone knows to set the POS parameters to 0. And if it's really to be applied everywhere it's better done within the parent class so a driver can't forget it (which then leads to a problem with the analog state needing to use it differently).

rmackay9 commented:
@WickedShell, I agree with you in many ways but I feel like we are trying to be too perfect in the long term but not helping in the short-term.  What we have now doesn't make much sense and I think this PR is an improvement.

I don't think this change really makes it harder to correct in the future.  A minor bit of param conversion and/or a warning to some users at the time of the release and it will be fine.  The larger coding effort of improving any estimator will far outweigh the addition of an offset parameter value to the distance.

WickedShell commented:
I think it's an impediment to the future in that all locations that read the value now need to remember to remove the offset to do it correctly. Which would lead to the sorta ridiculous situation where we add the offset first, then remove it later before applying a positional correction.

More critically though you can't actually remove that externally. Because you are using the value differently for the analog case from every other case you actually have to remove the offset only on the non analog driver case.

If the offset is purely used as a "ranging error correction" then I'm fine with it. But it really needs to never be used for mounting location correction as we won't be able to undo that in any well thought out manner in the future, short of making a new parameter and doing some complicated conversion into it, in order to fix the hack proposed now.

rmackay9 commented:
@WickedShell,

> I think it's an impediment to the future in that all locations that read the value now need to remember to remove the offset to do it correctly. Which would lead to the sorta ridiculous situation where we add the offset first, then remove it later before applying a positional correction.

No, we would never do this.

Your point about "ranging error correction" is a good one.  IMUs have both offsets and positions and this could be considered the same thing.

TBH, I need to move on so I've removed the 2nd commit and fixed the comment.  I'm merge after it passes travis.

rmackay9 commented:
Merged!

CI is failing but it seems to be because the firmware is slightly too big for one of the Matek boards which I think is fixed in master now.

rmackay9 commented:
FYI, this is included in Copter-4.0.2-rc1
